tests/models/test_infer_pip_requirements.py

```diff
@@ -0,0 +1,262 @@
+
+
+def _get_pinned_requirement(package_name):
+    return f"{package_name}=={importlib_metadata.version(package_name)}"
+
+
+def _read_requirements(model_uri):
+    reqs_uri = posixpath.join(model_uri, _REQUIREMENTS_FILE_NAME)
+    return _read_lines(_download_artifact_from_uri(reqs_uri))
+
+
+def _is_importable(module_name):
+    try:
+        __import__(module_name)
+        return True
+    except ImportError:
+        return False
+    except Exception:
+        _logger.exception("Encountered an unexpected error while importing `%s`", module_name)
+        return False
+
+
+def required_modules(*module_names):
+    """
+    Skips a test unless the specified modules are all available.
+    """
+
+    def decorator(test_func):
+        should_skip = not all(map(_is_importable, module_names))
+        reason = f"This test requires {module_names}"
+        return pytest.mark.skipif(should_skip, reason=reason)(test_func)
+
+    return decorator
+
+
+# REMOVE THIS FIXTURE BEFORE MERGING THE PR
+@pytest.fixture(autouse=True)
+def show_inferred_pip_requirements(request):
+    import functools
+
+    original = mlflow.infer_pip_requirements
+
+    @functools.wraps(original)
+    def patch(*args, **kwargs):
+        res = original(*args, **kwargs)
+
+        capture_manager = request.config.pluginmanager.getplugin("capturemanager")
+        capture_manager.suspendcapture()
+        title = "=" * 30 + " Inferred pip requirements " + "=" * 30
+        print("\n" + title)
+        print("\n".join(res))
+        print("=" * len(title))
+        capture_manager.resumecapture()
+
+        return res
+
+    with mock.patch("mlflow.infer_pip_requirements", new=patch):
+        yield
+
+
+@contextmanager
+def assert_did_not_fallback():
+    def side_effect(*args, **kwargs):
+        err_msg = (
+            "Encountered an unexpected error while inferring pip requirements "
+            "(model URI: %s, flavor: %s)"
+        )
+        if args[0] == err_msg:
+            raise Exception("SHOULD NOT FALLBACK")
+
+    with mock.patch("mlflow.utils.environment._logger.exception", side_effect=side_effect):
+        yield
+
+
+@pytest.fixture(params=range(2))
+def xgb_model(request):
+    import xgboost as xgb
+
+    model = xgb.XGBClassifier(objective="multi:softmax", n_estimators=10)
+    return [model, Pipeline([("model", model)])][request.param]
+
+
+@required_modules("xgboost")
+def test_infer_pip_requirements_xgboost(xgb_model):
+    X, y = load_iris(return_X_y=True, as_frame=True)
+    xgb_model.fit(X, y)
+
+    with mlflow.start_run(), assert_did_not_fallback():
+        mlflow.sklearn.log_model(xgb_model, artifact_path="model")
+        model_uri = mlflow.get_artifact_uri("model")
+
+    reqs = _read_requirements(model_uri)
+    assert _get_pinned_requirement("xgboost") in reqs
+    assert _get_pinned_requirement("scikit-learn") in reqs
+
+    data = X.head(3)
+    resp = pyfunc_serve_and_score_model(model_uri, data, CONTENT_TYPE_JSON_SPLIT_ORIENTED)
```

Comment:
Each of our model flavors have tests for model scoring, right? Can we incorporate this custom pipeline serving logic into those test modules instead?

Reply:
Got it!

Reply:
This is surprising but most of the flavors are missing serving tests.